src/lib/address_resolve/AddressResolve_DefaultImpl.cpp

```diff
@@ -0,0 +1,302 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/address_resolve/AddressResolve_DefaultImpl.h>
+
+namespace chip {
+namespace AddressResolve {
+namespace Impl {
+namespace {
+
+static constexpr System::Clock::Timeout kInvalidTimeout{ System::Clock::Timeout::max() };
+
+// IP addess "suitability"
+//   - Larger value means "more suitable"
+//   - Enum ordered ascending for easier read. Note however that order of
+//     checks MUST match in ScoreIpAddress below.
+enum class IpScore : unsigned
+{
+    kInvalid = 0, // No address available
+
+    // "Other" IPv6 include:
+    //   - invalid addresses (have seen router bugs during interop testing)
+    //   - embedded IPv4 (::/80)
+    kOtherIpv6                     = 1,
+    kIpv4                          = 2, // Not Matter SPEC, so low priority
+    kLinkLocal                     = 3, // Valid only on an interface
+    kUniqueLocal                   = 4, // ULA. Thread devices use this
```

Comment:
Unique local with shared prefix should exist and be before kGlobalUnicast. This is because global unicast may cause routing and external packets to be sent to the WAN, whereas ULAs will stay on the WAN. Matter is a LAN protocol, and all thread devices with ULAs (the core reason why you would see ULAs apart from special subnets like for dedicated IPV6 local DNS or other local subnets) are likely have to share prefix with the sender if there was no other *delegatable* address.

Reply:
Added. Thank you for noticing.